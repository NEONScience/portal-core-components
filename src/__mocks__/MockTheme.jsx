import React from 'react';

import CssBaseline from '@mui/material/CssBaseline';
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles';

import Theme from '../lib_components/components/Theme/Theme';

const MockTheme = ({ children }) => {
  return (
    <StyledEngineProvider injectFirst>
      <ThemeProvider theme={Theme}>
        <CssBaseline />
        {children}
      </ThemeProvider>
    </StyledEngineProvider>
  );
}

export default MockTheme;
