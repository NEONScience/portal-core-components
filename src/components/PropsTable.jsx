import React from 'react';
import PropTypes from 'prop-types';

import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';

import Theme, { COLORS } from '../lib_components/components/Theme/Theme';

const useStyles = makeStyles((theme) => ({
  propTableRowGrey: {
    backgroundColor: theme.palette.grey[50],
  },
  required: {
    marginTop: theme.spacing(2),
    fontSize: '0.9em',
    fontWeight: 600,
    color: COLORS.RED[400],
  },
}));

const PropsTable = (props) => {
  const { props: propRows, fullHeight } = props;
  const classes = useStyles(Theme);

  return (
    <TableContainer component={Paper} style={fullHeight ? {} : { maxHeight: '70vh' }}>
      <Table stickyHeader aria-label="props">
        <TableHead>
          <TableRow>
            <TableCell>Prop</TableCell>
            <TableCell>Type</TableCell>
            <TableCell>Default</TableCell>
            <TableCell>Examples</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {propRows.map((row, idx) => (
            <React.Fragment key={row.name}>
              <TableRow className={idx % 2 ? classes.propTableRowGrey : null}>
                <TableCell component="th" scope="row" rowSpan={2}>
                  <tt>{row.name}</tt>
                  {!row.required ? null : <div className={classes.required}>REQUIRED</div>}
                </TableCell>
                <TableCell>{row.type}</TableCell>
                <TableCell><tt>{row.default}</tt></TableCell>
                <TableCell>{row.examples ? row.examples : '--'}</TableCell>
              </TableRow>
              <TableRow className={idx % 2 ? classes.propTableRowGrey : null}>
                <TableCell colSpan={4}>{row.description}</TableCell>
              </TableRow>
            </React.Fragment>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

const stringOrNode = PropTypes.oneOfType([PropTypes.node, PropTypes.string]);
PropsTable.propTypes = {
  props: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      type: stringOrNode.isRequired,
      default: PropTypes.string,
      examples: stringOrNode,
      description: stringOrNode,
      required: PropTypes.bool,
    }),
  ).isRequired,
  fullHeight: PropTypes.bool,
};

PropsTable.defaultProps = {
  fullHeight: false,
};

export default PropsTable;
