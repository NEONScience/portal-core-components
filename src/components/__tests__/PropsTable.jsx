import React from 'react';
import renderer from 'react-test-renderer';

import MockTheme from '../../__mocks__/MockTheme';
import PropsTable from '../PropsTable';

describe('PropsTable', () => {
  test('renders a minimal single row', () => {
    const props = [
      {
        name: 'foo',
        type: 'string',
      },
    ];
    const tree = renderer.create(<MockTheme><PropsTable props={props} /></MockTheme>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders a multiple rows with optional strings', () => {
    const props = [
      {
        name: 'foo',
        type: 'string',
        default: 'abc',
      },
      {
        name: 'bar',
        type: 'string',
        description: 'lorem ipsum',
        examples: '"abc", "def"',
        required: true,
      },
    ];
    const tree = renderer.create(<MockTheme><PropsTable props={props} /></MockTheme>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders a multiple rows with optional nodes', () => {
    const props = [
      {
        name: 'foo',
        type: 'string',
        examples: <div>example node</div>,
      },
      {
        name: 'bar',
        type: <div>type node</div>,
        description: <div>description node</div>,
      },
    ];
    const tree = renderer.create(<MockTheme><PropsTable props={props} /></MockTheme>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders at full height', () => {
    const props = [
      {
        name: 'foo',
        type: 'string',
        description: 'lorem ipsum',
        examples: '"abc", "def"',
        required: true,
      },
      {
        name: 'bar',
        type: <div>type node</div>,
        description: <div>description node</div>,
      },
    ];
    const tree = renderer.create(<MockTheme><PropsTable props={props} fullHeight /></MockTheme>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
