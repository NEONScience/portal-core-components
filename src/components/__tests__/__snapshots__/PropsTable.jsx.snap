// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropsTable renders a minimal single row 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "maxHeight": "70vh",
    }
  }
>
  <table
    aria-label="props"
    className="MuiTable-root MuiTable-stickyHeader"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Prop
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Type
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Default
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Examples
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
          rowSpan={2}
          scope="row"
        >
          <tt>
            foo
          </tt>
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          string
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <tt />
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          --
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          colSpan={4}
        />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`PropsTable renders a multiple rows with optional nodes 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "maxHeight": "70vh",
    }
  }
>
  <table
    aria-label="props"
    className="MuiTable-root MuiTable-stickyHeader"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Prop
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Type
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Default
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Examples
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
          rowSpan={2}
          scope="row"
        >
          <tt>
            foo
          </tt>
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          string
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <tt />
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <div>
            example node
          </div>
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          colSpan={4}
        />
      </tr>
      <tr
        className="MuiTableRow-root makeStyles-propTableRowGrey-1"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
          rowSpan={2}
          scope="row"
        >
          <tt>
            bar
          </tt>
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <div>
            type node
          </div>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <tt />
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          --
        </td>
      </tr>
      <tr
        className="MuiTableRow-root makeStyles-propTableRowGrey-1"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          colSpan={4}
        >
          <div>
            description node
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`PropsTable renders a multiple rows with optional strings 1`] = `
<div
  className="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  style={
    Object {
      "maxHeight": "70vh",
    }
  }
>
  <table
    aria-label="props"
    className="MuiTable-root MuiTable-stickyHeader"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Prop
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Type
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Default
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader"
          scope="col"
        >
          Examples
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
          rowSpan={2}
          scope="row"
        >
          <tt>
            foo
          </tt>
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          string
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <tt>
            abc
          </tt>
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          --
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          colSpan={4}
        />
      </tr>
      <tr
        className="MuiTableRow-root makeStyles-propTableRowGrey-1"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          role="cell"
          rowSpan={2}
          scope="row"
        >
          <tt>
            bar
          </tt>
          <div
            className="makeStyles-required-2"
          >
            REQUIRED
          </div>
        </th>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          string
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          <tt />
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          "abc", "def"
        </td>
      </tr>
      <tr
        className="MuiTableRow-root makeStyles-propTableRowGrey-1"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
          colSpan={4}
        >
          lorem ipsum
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
